iroha_error/iroha_error_macro/src/error.rs

```diff
@@ -0,0 +1,55 @@
+extern crate proc_macro;
+
+use proc_macro_error::abort;
+use quote::quote;
+
+fn attrs_has_ident(attrs: &[syn::Attribute], ident: &str) -> bool {
+    attrs.iter().any(|attr| attr.path.is_ident(ident))
+}
+
+pub fn impl_source(ast: &syn::DeriveInput) -> proc_macro2::TokenStream {
+    let name = &ast.ident;
+
+    let variants = if let syn::Data::Enum(ref data_enum) = ast.data {
+        &data_enum.variants
+    } else {
+        abort!(ast, "Only enums are supported")
+    }
+    .iter()
+    .map(|variant| {
+        let impl_source = match variant.fields {
+            syn::Fields::Unnamed(ref fields) if fields.unnamed.len() == 1 && attrs_has_ident(&fields.unnamed[0].attrs, "from") => {
+                Some(true)
+            },
+            syn::Fields::Unnamed(ref fields) if fields.unnamed.len() == 1 => {
+                Some(false)
+            },
+            syn::Fields::Unit => None,
+
+            syn::Fields::Unnamed(_) => abort!(
+                variant,
+                "Unnamed variant should have exactly one arguments. Check out iroha2 style-guide."
+            ),
+            syn::Fields::Named(_) => abort!(
+                variant,
+                "Invalid variant. Named structures inside enum are not supported. Check out iroha2 style-guide."
+            ),
+        };
+        (&variant.ident, impl_source)
+    })
+    .map(|(variant, impl_source)| match impl_source {
+        Some(true) => quote! { Self:: #variant (var) => std::error::Error::source(var) },
+        Some(false) => quote! { Self:: #variant (_) => None },
+        None => quote! { Self:: #variant => None },
+    });
```

Comment:
I think two maps are not needed here, you can `quote` in the first, it will improve readability and reduce number of loc.